### ISSUE TITLE
--user --global available for more actions

### ISSUE THREAD

**Issue by es20490446e** [systemd/systemd#19284]
Currently, after package upgrade, there's no easy way to stop and then start again upgraded user daemons for all users, so they can get the new version running. That means users are required to log off to benefit from the upgrade.

I'm currently work-arounding this limitation with this [code](https://bin.privacytools.io/?62d0260dbeffddc1#5EtBtoUTqzC3aoFUEpgDvwAMShDZPWovTUz7n6QAbXRd), but this is rather a capability that systemd is missing.

systemd should be able to use "--user --global" with more actions like "start", "stop" and "daemon-reload".

Let me know if you will be eventually interested in implementing this capability. Otherwise, in three days, I will code a separate program to allow this.

**Comment by poettering** [comment]
No. Calling synchronously from privileged code into unprivileged code is not something we should ever supported. Moreover, what is this even supposed to do? Walk through all of /etc/passwd?

--global for enable/disable just means we operate on /etc/systemd/user/ instead of ~/.local/systemd/user. i.e. it has a minor change of directory as effect, and works strictly "offline", without talking to any daemons. But for calls suhc as "systemctl start" this doesn't really make sene, because it would mean talking to a myriad of service managers.

**Comment by es20490446e** [comment]
---
### Which users?

> Walk through all of /etc/passwd?

Can user units belong to users not listed by **loginctl**?

---
### What is prevented?

> Calling synchronously from privileged code into unprivileged code is not something we should ever supported.

Would it had been a better design of **root** if it was unable to perform certain tasks?

Will not supporting it **prevent** from doing it?

---
### How many managers?

> It would mean talking to a myriad of service managers.

How many **logged in** users could a system have?

What would be your suggested procedure for having a running unit **restarted** for all users, that is more direct than calling their service managers?

**Comment by poettering** [comment]
> Can user units belong to users not listed by **loginctl**?

Well you could, but that would be incomplete. i.e. think about stuff like "systemctl add-wants" and suchlike which certainly doesn't need a running session of a user to do something, but the users loginctl shows are just the logged in onces. So you'd only run this operation that would work perfectly fine for any user on those which happen to be logged in? how unfair?

> How many logged in users could a system have?

I think there's no boundary on that besides the max PID setting, which defaults to 4194304 these days. A bit much.

> What would be your suggested procedure for having a running unit restarted for all users, that is more direct than calling their service managers?

This was always an issue, and is unsolved. It would certainly already be better of the sessions listen to some event from the system if this is desirable. i.e. instead of having priv code to synchronously call into unpriv code, do it the other way round: have unpriv code asynchronously listen to some system event, that the privileged code fires off and forgets about.

That said, partial restarting of services after package database changes is always icky, since it's never comprehensive. it's not great on the system level, and it's not great on the user level either.

**Comment by es20490446e** [comment]
Thanks for the helping hand. This is what I have noticed during my life developing software:

Usually when the ideal solution is unclear it's better to implement something good enough. Then what is truly needed becomes self evident.

At the start of implementing a capability its design is just only a guess. By implementing the simplest case first, then iterating to include more cases, you get the true optimal design. As on each iteration you can contrast with reality.

You no longer need to figure out all that ideal design in you head. You simply iterate over the design till you cover most cases.

Anything extra can be included later on, just as needed. And usually that is never.

Cheers.

https://youtu.be/z59w6HHpS2w

**Comment by es20490446e** [comment]
This is the stuff I enjoy reading the most:
https://cs.stanford.edu/people/eroberts/courses/cs181/projects/2010-11/WorseIsBetter/index.php/Worse-is-better.html

**Comment by jo-so** [comment]
Restarting user units after an upgrade might be a good job for [needrestart](https://github.com/liske/needrestart). Even unattended restart of system units should be done with a tool that does not complex checking/exclusion. For example, I start my emacs and firefox with a user unit and I wouldn't be happy if they become automatically restarted on update.

**Comment by es20490446e** [comment]
Why not just restarting from a package script? Wouldn't that provide better control over what to restart?

**Comment by jo-so** [comment]
The main problem are library updates. The library package doesn't know which services are using the library and need a restart to get the new version. And the service package doesn't know of the library package updates.

The second problem are double/triple restarts. Some services would be restarted multiple times due to dependency updates. Which is not ideal and takes a long time in the worst case. Dpkg invented *trigger* for this to tell the package manager to run a special update at the end. One heavy example was texlive that needed a rebuild of a database which happens with triggers only once, but multiple times before.

A third problem are services you can't automatically restart, but under some conditions you can. One example is dbus. On a desktop this kills your Gnome session, but on a web server you can.

Unfortunately, the task of service restarts sound simple, but is not that easy.

**Comment by es20490446e** [comment]
What are the consequences of not restarting the services at all during library updates?

That the libraries on RAM will be the older versions till next restart? Or there will be any other consequence?

**Comment by es20490446e** [comment]
Thanks for the info 👍
